AnimationGroup.lua

```diff
@@ -60,28 +60,49 @@ function CreateFrame(...)
     return frame
 end
 
-
 local function OnUpdate(self, elapsed)
     if self.paused then
         return
     end
-
-    self.time = self.time + (self.group.reverse and -elapsed or elapsed)
+    self.delaying = true
+
+    if not self.group.reverse and self.startdelay and self.startdelayTime < self.startdelay then
+        self.startdelayTime = self.startdelayTime + elapsed
+    elseif self.group.reverse and self.enddelay and self.enddelayTime < self.enddelay then
+        self.enddelayTime = self.enddelayTime + elapsed
+    elseif (not self.group.reverse and self.time < self.duration) or (self.group.reverse and self.time > 0) then
+        self.time = self.time + (self.group.reverse and -elapsed or elapsed)
+        self.delaying = false
+        self.progress = self.time / self.duration
+        self.progress = self.progress < 0 and 0 or self.progress
+        self.progress = self.progress > 1 and 1 or self.progress
+        self.smoothProgress = self.smoothing_func(self.progress).y
+    end
```

Comment:
There's a lot going on here. I'd encourage you to assign the conditions to variables that indicate what they test, e.g.:

```
local is_progressing_ = not self.group.reverse and self.time < self.duration
local is_progressing_reverse = self.group.reverse and 0 < self.time
```

Note, that I also flipped `0 < self.time`. Reading left to right is easier on the brain :) (good pattern to always apply, `<` rather than `>`).

Reply:
reworked